src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,693 @@
+
+  \todo handle the unlinking of the admin socket. Note that 'cleanup_files'
+  at-exit functionality is not yet implemented in Crimson.
+*/
+
+namespace {
+seastar::logger& logger()
+{
+  return crimson::get_logger(ceph_subsys_osd);
+}
+}  // namespace
+
+namespace crimson::admin {
+
+/**
+  Hooks table - iterator support
+ */
+
+AdminHooksIter AdminSocket::begin()
+{
+  AdminHooksIter it{ *this };
+  return it;
+}
+
+AdminHooksIter AdminSocket::end()
+{
+  AdminHooksIter it{ *this, true };
+  return it;
+}
+
+/**
+  note that 'end-flag' is an optional parameter, and is only used internally (to
+  signal the end() state).
+ */
+AdminHooksIter::AdminHooksIter(AdminSocket& master, bool end_flag)
+    : m_master{ master }, m_end_marker{ end_flag }
+{
+  if (end_flag) {
+    // create the equivalent of servers.end();
+    m_miter = m_master.servers.end();
+  } else {
+    m_miter = m_master.servers.begin();
+    m_siter = m_miter->second.m_hooks.begin();
+  }
+}
+
+AdminHooksIter& AdminHooksIter::operator++()
+{
+  if (!m_end_marker) {
+    ++m_siter;
+    if (m_siter == m_miter->second.m_hooks.end()) {
+      // move to the next server-block
+      m_miter++;
+      if (m_miter == m_master.servers.end()) {
+        m_end_marker = true;
+        return *this;
+      }
+
+      m_siter = m_miter->second.m_hooks.begin();
+    }
+  }
+  return *this;
+}
+
+/**
+  the default implementation of the hook API
+
+  Note that we never throw or return a failed future.
+ */
+seastar::future<bool> AdminSocketHook::call(std::string_view  command,
+                                            const cmdmap_t&   cmdmap,
+                                            std::string_view  format,
+                                            ceph::bufferlist& out) const
+{
+  unique_ptr<Formatter> f{ Formatter::create(format, "json-pretty"sv,
+                                             "json-pretty"s) };
+
+  // customize the output section, as required by a couple of hook APIs:
+  std::string section{ section_name(command) };
+  boost::replace_all(section, " ", "_");
+  if (format_as_array()) {
+    f->open_array_section(section.c_str());
+  } else {
+    f->open_object_section(section.c_str());
+  }
+
+  bool bres{ false };  ///< 'false' for any failure of in API execution
+
+  /*
+    call the command-specific hook.
+    A note re error handling:
+        - will be modified to use the new 'erroretor'. For now:
+        - exec_command() may throw or return an exceptional future. We return an
+           error message on both failure scenarios.
+   */
+  return seastar::do_with(
+    std::move(f), std::move(bres),
+    [this, &command, &cmdmap, &format, &out](unique_ptr<Formatter>& ftr,
+                                             bool&                  br) {
+      return seastar::futurize_apply(
+               [this, &command, &cmdmap, &format, &out, f = ftr.get()] {
+                 return exec_command(f, command, cmdmap, format, out);
+               })
```

Comment:
What's the advantage of using futurize_apply here rather than simply calling exec_command directly?

Reply:
AdminSocket::call() does not, by design, throw or fut-throw. Thus I must handle the cases of exec_command() either throwing or returning a failed future. (This is why I've added the 'throw', 'fthrow', ... etc' asok commands - to be able to test the correct behaviour if exec_command() does not succeed).
Using futurize_apply() allows me to:
- avoid explicit try-catch here, and
- code just one error path instead of two.